src/modules/flow/twitter/twitter.c

```diff
@@ -0,0 +1,617 @@
+        "&oauth_token=%s"
+        "&oauth_version=1.0",
+        mdata->consumer_key, cb_data->nonce,
+        cb_data->timestamp, mdata->token);
+    SOL_INT_CHECK_GOTO(r, < 0, err_params);
+
+    r = sol_http_encode_slice(&buf, sol_str_slice_from_str(params));
+    SOL_INT_CHECK_GOTO(r, < 0, err_escape_params);
+    escaped_params = sol_str_slice_to_string(sol_buffer_get_slice(&buf));
+    sol_buffer_fini(&buf);
+    SOL_NULL_CHECK_GOTO(escaped_params, err_escape_params);
+
+    r = asprintf(&signature, "GET&%s&%s", mdata->escaped_get_url, escaped_params);
+    SOL_INT_CHECK_GOTO(r, < 0, err_signature);
+
+    blob = sol_blob_new(SOL_BLOB_TYPE_DEFAULT, NULL, signature, strlen(signature));
+    SOL_NULL_CHECK_GOTO(blob, err_blob);
+
+    r = sol_message_digest_feed(digest, blob, true);
+    SOL_INT_CHECK_GOTO(r, < 0, err_feed);
+
+    sol_blob_unref(blob);
+    free(escaped_params);
+    free(params);
+
+    return 0;
+
+err_feed:
+    sol_blob_unref(blob);
+err_blob:
+    free(signature);
+err_signature:
+    free(escaped_params);
+err_escape_params:
+    free(params);
+err_params:
+    sol_message_digest_del(digest);
+err_digest:
+    free(cb_data->key);
+err_key:
+    free(cb_data->timestamp);
+err_timestamp:
+    free(cb_data->nonce);
+err:
+    free(cb_data);
+    return r;
+}
+
+static int
+post_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id,
+    const struct sol_flow_packet *packet)
+{
+    int r;
+    const char *val;
+    struct twitter_data *mdata = data;
+
+    if (!(mdata->token && mdata->token_secret)) {
+        sol_flow_send_error_packet(node, EINVAL,
+            "There is no access token");
+        return -EINVAL;
+    }
+
+    r = sol_flow_packet_get_string(packet, &val);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return post_status(node, val);
+}
+
+static void
+twitter_close(struct sol_flow_node *node, void *data)
+{
+    uint16_t i;
+    struct twitter_data *mdata = data;
+    struct sol_http_client_connection *connection;
+
+    free(mdata->consumer_key);
+    free(mdata->consumer_secret);
+    free(mdata->token);
+    free(mdata->token_secret);
+    free(mdata->escaped_post_url);
+    free(mdata->escaped_get_url);
+    SOL_PTR_VECTOR_FOREACH_IDX (&mdata->pending_conns, connection, i)
+        sol_http_client_connection_cancel(connection);
+}
+
+static int
+twitter_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    int r;
+    struct twitter_data *mdata = data;
+    struct sol_buffer buf;
+    struct sol_flow_node_type_twitter_client_options *opts =
+        (struct sol_flow_node_type_twitter_client_options *)options;
+
+    mdata->consumer_key = strdup(opts->consumer_key);
+    SOL_NULL_CHECK(mdata->consumer_key, -ENOMEM);
+
+    r = sol_http_encode_slice(&buf, sol_str_slice_from_str(opts->consumer_secret));
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+    mdata->consumer_secret = sol_str_slice_to_string(sol_buffer_get_slice(&buf));
```

Comment:
look at your code:
 - if encoded, then we guarantee a null byte
 - if not encoded, it's the same buffer you gave it, that is opts->consumer_secret, that is known to be null-terminated, after all you created the slice from that string using the null terminator.

You can keep like that since it's the open() node and happens only once, also because you need to dup opts->consumer_secret since we shouldn't keep references to opts pointers. But consider this for other places